cre.cpp

```diff
@@ -1510,28 +1519,676 @@ static int getPageLinks(lua_State *L) {
+                    extStopReason = "page-break-before met";
+                    break;
+                }
+                if ( pb_after == css_pb_always || pb_after == css_pb_left || pb_after == css_pb_right ) {
+                    ldomXPointerEx tmpPos = curPos;
+                    // printf("[pbafter at %s\n", UnicodeToLocal(notAfter.toString()).c_str());
+                    if ( tmpPos.nextOuterElement() ) {
+                        notAfter = tmpPos;
+                        // printf("[notAfter %s\n", UnicodeToLocal(notAfter.toString()).c_str());
+                    }
+                }
+                // When we meet another final block containing a node with an ID= attribute,
+                // it's probably another footnote.
+                // (In the first final block, it's possible to have multiple nodes with
+                // different ID, which could mean there are multiple terms or synonyms...)
+                lString16 id = node->getAttributeValue("id");
+                if ( !id.empty() ) {
+                    // printf("id=%s\n", UnicodeToLocal(id).c_str());
+                    extStopReason = "node with 'id=' attr met";
+                    break;
+                }
+                // Done checking
+                if ( !curPos.nextElement() ) {
+                    extStopReason = "end of document met";
+                    break;
+                }
+                // printf("[...: %s\n", UnicodeToLocal(curPos.toString()).c_str());
+                if ( !notAfter.isNull() && curPos.compare(notAfter) >= 0 ) {
+                    extStopReason = "page-break-after met";
+                    if ( curFinalNode && !curFinalNode->isNull() )
+                        goodFinalNode = curFinalNode;
+                    break;
+                }
+            } // end of while (true)
+            extendedStopReason = extStopReason;
+            if ( goodFinalNode && !goodFinalNode->isNull() ) {
+                // printf("GOOD final node\n");
+                ldomXPointerEx extendedEnd = ldomXPointerEx(goodFinalNode, 0);
+                extendedEnd.lastInnerNode(true); // We may miss a trailing image
+                extendedRange = ldomXRange(extendedStart, extendedEnd);
+            }
+            // If we'd get multiple <LI> in multiple final nodes, we'll
+            // be requesting the HTML from the common parent, and if it
+            // is a <OL>, we'll start numbering the <LI> from "1", which
+            // may not be the original numbering in the document.
+            // Howerver, this should not happen with documents using <LI id=...>
+            // as the container for each footnote (like Wikipedia EPUBs), as
+            // the ID= check will then prevent us from including multiple <LI>
+            // in our extended range: the root node will be the <LI>, that we
+            // will mask with CSS "body > li { list-style-type: none; }".
+        }
+    }
+
+    // Target text must be < 10000 chars
```

Comment:
I wonder what the amount of characters is on some of those academic footnotes…

Also, I hope we're talking displayed text right? Because I've seen some `<span><span><span><span><span><span><span><span><span><span><span><span><span><span>tiny bit of text` type stuff around. ;-)

Reply:
> I hope we're talking displayed text right?

Yes we do :) That's the point of this function, otherwise I could have just used `#html` on the lua side.
This just counts text length text node by text node. I could have just used topNode:getText() to get the full text and get its length, but I just made this function so it stops counting when limit is reached, so we don't have a whole book text in memory before we decide its too much.

MuPDF can cope fine with huge text, but it's just to not block the use too long while MuPDF renders.
5000 is what kobo chose, I met one footnote of 6000, so I pushed it to 10000 just be be large.
But I think I'll make that SIZE_LIMIT passed from Lua, just so we can tweak it from there if need be.

Reply:
5000 seems rather small when I've seen a three page footnote in a minuscule footnote font.

(This was in a German book. Those can be something else.)

Reply:
Well, that's also the point of this test: when it's that long, it's probably more confortable to go read it in the book (false => follow link) instead of in that 1/3 screen.height footnote widget.